### PR TITLE
Ingest byw

### PR DESCRIPTION
**Pull request by LishaRamon**
Ingesting the sources from Rothermich 2024

Link to relevant issue: Closes # 428

For data ingests:
- [x] includes script used for ingest
- [x] includes modified JSON files
- [x] Add new tests
- [ ] ~~Update the Versions table~~

### CODE REVIEW

**Comment by LishaRamon** [comment]
Line 17 Error: "astropy.io.ascii.core.InconsistentTableError: Number of header columns (9) inconsistent with data columns (7) at data line 0"

Header values: ['Source', 'RA', 'Dec', 'Epoch', 'Equinox', 'Shortname', 'Reference', 'Other_ref', 'Comments']
Data values: ['CWISE J000021.45-481314.9', '0.0893808', '-48.2208077', '2015.4041', 'ICRS', 'CWISE 0000-4813', 'Rothermich et al. Submitted']

I understand how it's seeing there's missing data under two of the header columns but not sure how to go about debugging for it to recognize that having a missing value is okay.

File: scripts/ingests/ingest_BYW_2024.py [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+
+data_columns = ['CWISE J000021.45-481314.9', '0.0893808', '-48.2208077', '2015.4041', 'ICRS', 'CWISE 0000-4813', 'Rothermich', 'None', 'None']  # columns with wanted data values
+
+print(byw_table)
+
+# replacing empty values ('cdots') with None
+for column in data_columns:
+    byw_table[column][np.where(byw_table[column] == 'cdots')] = None
+
+def ingest_source(db, source, reference=None, ra=None, dec=None, epoch=None, equinox=None,
```
Comment (kelle) [inline scripts/ingests/ingest_BYW_2024.py:28]: This function is already defined in ingest_utils and imported. You just have to **use** the function. (Same as you used ingest_proper_motions.) Delete lines 28 onward.

**Comment by kelle** [comment]
I think you should use astropy.io.ascii instead of astropy.Table. It looks like ascii.read does a better job of handling empty cells:

https://docs.astropy.org/en/stable/io/ascii/read.html

**Comment by LishaRamon** [comment]
Line 17 Error Update:
ValueError: ASCII format 'ascii' not in allowed list ['aastex', 'basic', 'cds', 'commented_header', 'csv', 'daophot', 'ecsv', 'fast_basic', 'fast_commented_header', 'fast_csv', 'fast_no_header', 'fast_rdb', 'fast_tab', 'fixed_width', 'fixed_width_no_header', 'fixed_width_two_line', 'html', 'ipac', 'latex', 'mrt', 'no_header', 'qdp', 'rdb', 'rst', 'sextractor', 'tab']

How would I find/recognize the correct 'format'? I assumed csv since it's an excel file, but it breaks the code further.

File: scripts/ingests/ingest_BYW_2024.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+RECREATE_DB = True  # recreates the .db file from the data files
+# LOAD THE DATABASE
+db = load_simpledb('SIMPLE.db', recreatedb=RECREATE_DB)
+
+
+# live google sheet
+link = "https://docs.google.com/spreadsheets/d/1JFa8F4Ngzp3qAW8NOBurkz4bMKo9zXYeF6N1vMtqDZs/edit#gid=0"
+
+columns = ['source', 'ra', 'dec', 'epoch', 'equinox', 'shortname', 'reference', 'other_ref', 'comments']
+byw_table = ascii.read(link, format='ascii', data_start=2, data_end=90, header_start=1, names=columns, guess=False,
```
Comment (kelle) [inline scripts/ingests/ingest_BYW_2024.py:17]:
```suggestion
byw_table = ascii.read(link, format='csv', data_start=2, data_end=90, header_start=1, names=columns, guess=False,
```

File: scripts/ingests/ingest_BYW_2024.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+                           fast_reader=False, delimiter=',')
+
+data_columns = ['Source', 'RA', 'Dec', 'Epoch', 'Equinox', 'Shortname', 'Reference', 'Other_ref', 'Comments']  # columns with wanted data values
+
+# replacing empty values ('cdots') with None
+for column in data_columns:
+    byw_table[column][np.where(byw_table[column] == 'cdots')] = None
+
```
Comment (kelle) [inline scripts/ingests/ingest_BYW_2024.py:25]: You can delete this. You don't have any cdots that need replacing

File: scripts/ingests/ingest_BYW_2024.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+from astropy.coordinates import Angle
+
+SAVE_DB = True  # save the data files in addition to modifying the .db file
+RECREATE_DB = True  # recreates the .db file from the data files
+# LOAD THE DATABASE
+db = load_simpledb('SIMPLE.db', recreatedb=RECREATE_DB)
+
+
+# live google sheet
+link = "https://docs.google.com/spreadsheets/d/1JFa8F4Ngzp3qAW8NOBurkz4bMKo9zXYeF6N1vMtqDZs/edit#gid=0"
```
Comment (kelle) [inline scripts/ingests/ingest_BYW_2024.py:14]: The problem is here: you need to change the end of this link to something different. It shouldn't be edit. Look at allys script to find the correct suffix. 
```suggestion
link = "https://docs.google.com/spreadsheets/d/1JFa8F4Ngzp3qAW8NOBurkz4bMKo9zXYeF6N1vMtqDZs/<something different>"
```

Comment (LishaRamon) [reply scripts/ingests/ingest_BYW_2024.py]: Originally I had "pub?output=csv" at the end  of the link but it gave HTTP Error 404 Unauthorized, but keeping it as it's original link seemed to resolve the issue.

File: scripts/ingests/ingest_BYW_2024.py [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+link = "https://docs.google.com/spreadsheets/d/1JFa8F4Ngzp3qAW8NOBurkz4bMKo9zXYeF6N1vMtqDZs/pub?output=csv"
+
+columns = ['source', 'ra', 'dec', 'epoch', 'equinox', 'shortname', 'reference', 'other_ref', 'comments']
+byw_table = ascii.read(link, format='ecsv', data_start=2, data_end=90, header_start=1, names=columns, guess=False,
+                           fast_reader=False, delimiter=',')
+
+data_columns = ['Source', 'RA', 'Dec', 'Epoch', 'Equinox', 'Shortname', 'Reference', 'Other_ref', 'Comments']  # columns with wanted data values
+
+
+def ingest_source(db):
```
Comment (kelle) [inline scripts/ingests/ingest_BYW_2024.py:23]: Careful. A function already exists with this name. You imported it from utils. This function should be named something different.

File: scripts/ingests/ingest_BYW_2024.py [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+    ingest_source(db, source = ["CWISE J000021.45-481314.9"], 
+                  reference = "Rothermich", 
+                  ra = [0.0893808], 
+                  dec = [-48.2208077], 
+                  epoch = [2015.4041], 
+                  equinox = "ICRS")
```
Comment (kelle) [inline scripts/ingests/ingest_BYW_2024.py:30]: Good idea to try this! Does it work?

**Comment by kelle** [comment]
If you can't figure out how to read the Google sheet, just use the CSV file from the issue.

File: scripts/ingests/ingest_BYW_2024.py [hunk 7]
```diff
@@ -11,24 +12,31 @@
 
 
 # live google sheet
-link = "https://docs.google.com/spreadsheets/d/1JFa8F4Ngzp3qAW8NOBurkz4bMKo9zXYeF6N1vMtqDZs/pub?output=csv"
+sheet_id = '1JFa8F4Ngzp3qAW8NOBurkz4bMKo9zXYeF6N1vMtqDZs'
 
-columns = ['source', 'ra', 'dec', 'epoch', 'equinox', 'shortname', 'reference', 'other_ref', 'comments']
-byw_table = ascii.read(link, format='ecsv', data_start=2, data_end=90, header_start=1, names=columns, guess=False,
-                           fast_reader=False, delimiter=',')
+df = pd.read_csv(f"https://docs.google.com/spreadsheets/d/{sheet_id}/export?format=csv")
```
Comment (kelle) [inline scripts/ingests/ingest_BYW_2024.py:17]: I said specifically that you don't need the pandas part, just the help with formatting the link.
Use this string as your link: f"https://docs.google.com/spreadsheets/d/{sheet_id}/export?format=csv"
or whatever you learned in the video.  Notice the difference between what you had before:
`pub?output=csv` and what you have now: `pub?output=csv`

**Comment by LishaRamon** [comment]
Since ascii.read attempted to read data from local files rather from URLs, by using a library like 'requests' it helped get data and create an object that can be passed to ascii.read to understand

**Comment by kelle** [comment]
Great work!

File: data/cwise_j000021.45-481314.9.json [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+        {
+            "source": "CWISE J000021.45-481314.9",
+            "ra": 0.0893808,
+            "dec": -48.2208077,
+            "epoch": 2015.4041,
+            "equinox": "ICRS",
+            "shortname": null,
+            "reference": "Roth",
+            "other_references": "\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000",
+            "comments": "\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000"
```
Comment (kelle) [inline data/cwise_j000021.45-481314.9.json:12]: Hmmmm, these characters are not ideal.

File: scripts/ingests/ingest_BYW_2024.py [hunk 9]
```diff
@@ -1,108 +1,69 @@
...
+
+        for col_name in row.colnames:
+            print(f"{col_name}: {row[col_name]}")
+             
+ 
+        ingest_source(db, source = row['Source'], 
+                    reference= row['Reference'], 
+                    ra= row['RA'], 
+                    dec= row['Dec'], 
+                    comment= row['Comments'],
```
Comment (kelle) [inline scripts/ingests/ingest_BYW_2024.py:39]: There are no comments, so don't ingest any.
```suggestion
                     
```

File: scripts/ingests/ingest_BYW_2024.py [hunk 10]
```diff
@@ -1,108 +1,69 @@
...
+             
+ 
+        ingest_source(db, source = row['Source'], 
+                    reference= row['Reference'], 
+                    ra= row['RA'], 
+                    dec= row['Dec'], 
+                    comment= row['Comments'], 
+                    epoch= row['Epoch'],
+                    equinox= row['Equinox'], 
+                    other_reference= row['Other_ref'],
```
Comment (kelle) [inline scripts/ingests/ingest_BYW_2024.py:42]: Since these are all blank, I would just omit this.
```suggestion
                    
```

**Comment by LishaRamon** [comment]
Not sure why the json files didnt update when I added the changes, but they didn't show up for me to push new ones

**Review by kelle — APPROVED** [review]
This looks great to me. I think we can merge this tomorrow.

File: data/cwise_j000021.45-481314.9.json [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+            "equinox": "ICRS",
+            "shortname": null,
+            "reference": "Roth",
+            "other_references": null,
+            "comments": null
+        }
+    ],
+    "Names": [
+        {
+            "other_name": "CWISE J000021.45-481314.9"
```
Comment (kelle) [inline data/cwise_j000021.45-481314.9.json:17]: Missing short name

File: scripts/ingests/ingest_backyard_worlds_benchmarks.py [hunk 12]
```diff
@@ -0,0 +1,3 @@
+test
+
```
Comment (kelle) [inline scripts/ingests/ingest_backyard_worlds_benchmarks.py:2]: This file shouldn't be part of the ingest. I would just delete the entire file.

**Review by kelle — APPROVED** [review]

**Pull request merged**